rtmp-server/src/amf0/parser.rs

```diff
@@ -0,0 +1,145 @@
+    InsufficientData,
+    #[error("Invalid UTF-8 string")]
+    InvalidUtf8,
+}
+
+const OBJECT_END_MARKER: [u8; 3] = [0x00, 0x00, 0x09];
+
+pub struct Parser;
+
+impl Parser {
+    pub fn parse(data: &[u8]) -> Result<Vec<AmfValue>, ParseError> {
+        let mut buf = Bytes::copy_from_slice(data);
+        let mut result = Vec::new();
+
+        while buf.has_remaining() {
+            let (value, remaining) = Self::parse_value(buf)?;
+            result.push(value);
+            buf = remaining;
+        }
+
+        Ok(result)
+    }
+
+    fn parse_value(mut buf: Bytes) -> Result<(AmfValue, Bytes), ParseError> {
+        if !buf.has_remaining() {
+            return Err(ParseError::InsufficientData);
+        }
+
+        let marker = buf.get_u8();
+
+        match marker {
+            // number
+            0x00 => {
+                if buf.remaining() < 8 {
+                    return Err(ParseError::InsufficientData);
+                }
+                let number = buf.get_f64();
+                Ok((AmfValue::Number(number), buf))
+            }
+            // bool
+            0x01 => {
+                if buf.remaining() < 1 {
+                    return Err(ParseError::InsufficientData);
+                }
+                let boolean = buf.get_u8() == 1;
+                Ok((AmfValue::Boolean(boolean), buf))
+            }
+            // string
+            0x02 => {
+                if buf.remaining() < 2 {
+                    return Err(ParseError::InsufficientData);
+                }
+                let size = buf.get_u16() as usize;
+                if buf.remaining() < size {
+                    return Err(ParseError::InsufficientData);
+                }
+                let string_bytes = buf.copy_to_bytes(size);
+                let string = String::from_utf8(string_bytes.to_vec())
+                    .map_err(|_| ParseError::InvalidUtf8)?;
+                Ok((AmfValue::String(string), buf))
+            }
+            // object
+            0x03 => {
+                let (pairs, remaining) = Self::parse_object_pairs(buf)?;
+                Ok((AmfValue::Object(pairs), remaining))
+            }
+            // null
+            0x05 => Ok((AmfValue::Null, buf)),
+            // ECMA array
+            0x08 => {
+                if buf.remaining() < 4 {
+                    return Err(ParseError::InsufficientData);
+                }
+                let _array_size = buf.get_u32();
+                let (pairs, remaining) = Self::parse_object_pairs(buf)?;
+                Ok((AmfValue::EcmaArray(pairs), remaining))
+            }
+            // strict array
+            0x0A => {
+                if buf.remaining() < 4 {
+                    return Err(ParseError::InsufficientData);
+                }
+                let size = buf.get_u32() as usize;
+                let mut array = Vec::with_capacity(size);
+                let mut current_buf = buf;
+
+                for _ in 0..size {
```

Comment:
You don't know that this message is in the buffer, you only checked if you have at least 4 bytes